File: SemanticSearch/Java/Web/SpringBoot/src/main/java/com/portfolio/semanticsearch/vectorstore/VectorStoreAdapter.java
package com.portfolio.semanticsearch.vectorstore;

import java.util.List;
import java.util.Map;

public interface VectorStoreAdapter {
    void connect();
    void addDocuments(List<String> documents, List<float[]> embeddings, List<Map<String, Object>> metadata);
    List<SearchResult> search(float[] queryEmbedding, int nResults);
    void deleteCollection(String collectionName);
    List<String> listCollections();
    void close();
}

File: SemanticSearch/Elixir/Web/Phoenix/src/lib/semantic_search/vectorstore/vector_store_adapter.ex
defmodule SemanticSearch.VectorStore.VectorStoreAdapter do
  @moduledoc """
  Behaviour for vector store adapters (ChromaDB, PgVector, Pinecone, InMemory).
  """

  @callback connect() :: :ok | {:error, term()}
  @callback add_documents([String.t()], [list(number())], [map()]) :: :ok | {:error, term()}
  @callback search(list(number()), pos_integer()) :: [SemanticSearch.VectorStore.SearchResult.t()]
  @callback delete_collection(String.t()) :: :ok | {:error, term()}
  @callback list_collections() :: [String.t()]
end
